GM: Buick LaCrosse 2017-19 support

koch-cf opened the pull request:
**Checklist**
- [ ] added entry to CarInfo in selfdrive/car/*/values.py and ran `selfdrive/car/docs.py` to generate new docs
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with openpilot:
- [ ] route with stock system:

**Questions**
1.  I'm still learning Python and Git, what's the best way to run docs.py to generate the new docs?
2. What criteria does the test route need to meet?

koch-cf committed:
Added Buick LaCrosse 2017

koch-cf committed:
Added Buick LaCrosse as candidate

koch-cf committed:
Added Buick LaCrosse CAR_INFO and Fingerprint

koch-cf committed:
Added Buick LaCrosse to non_tested_cars

koch-cf committed:
Added Buick LaCrosse

koch-cf committed:
Updated number of supported cars to 237

koch-cf committed:
Added ACC and LKAS description to Buick LaCrosse

sshane commented on selfdrive/car/gm/values.py:
```diff
@@ -98,6 +99,7 @@ def init_make(self, CP: car.CarParams):
   CAR.CADILLAC_ATS: GMCarInfo("Cadillac ATS Premium Performance 2018"),
   CAR.MALIBU: GMCarInfo("Chevrolet Malibu Premier 2017"),
   CAR.ACADIA: GMCarInfo("GMC Acadia 2018", video_link="https://www.youtube.com/watch?v=0ZN6DdsBUZo"),
+  CAR.BUICK_LACROSSE: GMCarInfo("Buick LaCrosse Premium 2017", "Adaptive Cruise Control (ACC) & LKAS"),
```
Does it just support 2017, or should go all the way to 2019 like you put in CARS.md?

sshane replied:
If you have Adaptive Cruise Contro, LKAS can still be a separate package?

koch-cf replied:
It’s probably 2017-19. According to the window sticker LKAS was standard equipment for the Premium but ACC was an option.

koch-cf replied:
Changed to 2017-19 and removed LKAS description

sshane commented on selfdrive/car/tests/routes.py:
```diff
@@ -19,6 +19,7 @@
   FORD.ESCAPE_MK4,
   FORD.FOCUS_MK4,
   FORD.MAVERICK_MK1,
+  GM.BUICK_LACROSSE,
```
Can you get a route on this branch, keeping it mind it will be public and for CI testing? We usually recommend starting the route/car in a parking lot, then drive around engaged for a few minutes, and ending in a parking lot as well.

koch-cf replied:
I'm still not able to have this branch recognize my car as "supported", and it goes into dashcam mode.  It does not seem to be the fault of the fingerprint, as I can move the fingerprint under the Malibu and it then thinks that my car is a Malibu and will work.  So there must be something else wrong with my modifications but I've not been able to find what it is.

sshane replied:
Fixed, can you unprotect your master branch so I can push to it and resolve the merge conflict?

koch-cf replied:
I have removed the branch protection rule, let me know if anything else is needed

koch-cf replied:
I took a short drive with the current code and got a radar fault. The route is 2dfdef71f033f267|2023-02-16--07-00-02

sshane replied:
Can you put your device on WiFi to upload logs? It sounds like there may be a communication issue somewhere, usually a bad cable.

sshane replied:
If you can get a route on this branch now without any faults, then I think we're ready to merge! Again a reminder that this route will be public, so a drive from parking lot, engaged, to parking lot is ideal.

koch-cf replied:
> Can you put your device on WiFi to upload logs? It sounds like there may be a communication issue somewhere, usually a bad cable.

I had the device on Wifi for a few hours this afternoon so hopefully they uploaded. If not it’ll be tonight.

koch-cf replied:
> If you can get a route on this branch now without any faults, then I think we're ready to merge! Again a reminder that this route will be public, so a drive from parking lot, engaged, to parking lot is ideal.

I’m getting way too many radar faults to be able to make a clean route.

Verylukyguy replied:
Make sure that all of your connections are good, even your add-a-circuit that powers the radar module.

sshane replied:
Did you upload rlogs for a route @koch-cf? I can check it out briefly

koch-cf replied:
I ran a test route last night with the new hardware (new C3 and new OBDII harness).  Still getting radar errors.  I've uploaded all of the files, note that this is a new dongle ID since it's new hardware:  75a6bcb9b8b40373|2023-02-25--00-12-54 .  I need some help diagnosing where these errors are coming from.  

Also note that the longitudinal tuning is also still very poor.  The car would have rear-ended the car at 19:22:30 if I had not applied the brakes at the last minute.  Both acceleration and braking are very slow, to the point where I often need to manually intervene.

sshane replied:
Looks like the radar messages on bus 1 are coming in inconsistently, and we're dropping messages. Total error count for the panda for can 1 is also rising

<img width="1118" alt="Screenshot_46" src="https://user-images.githubusercontent.com/25857203/221386531-97c9563d-9bd2-4a07-9e10-41975719d01b.png">

koch-cf replied:
Fixed radar error by removing the 1ft extension cable and uncoiling the USB-C cable under the dash.  Added test route 75a6bcb9b8b40373|2023-03-11--22-47-33 .

sshane commented:
Can you post your dongle ID as well?

koch-cf committed:
Updated CAR_INFO for Buick LaCrosse

koch-cf committed:
Added the Escalade which was recently updated

koch-cf commented:
> Can you post your dongle ID as well?

Dongle ID 2dfdef71f033f267

sshane commented on selfdrive/car/gm/values.py:
```diff
@@ -150,6 +152,11 @@ class CanBus:
+  CAR.BUICK_LACROSSE: [
+  # LaCrosse Premium AWD 2017
+  {
+    190: 6, 192: 5, 193: 8, 197: 8, 199: 4, 201: 6, 209: 7, 211: 2, 241: 6, 249: 8, 289: 1, 290: 1, 298: 2, 304: 8, 309: 8, 311: 8, 313: 8, 320: 8, 322: 7, 328: 1, 352: 7, 353: 3, 368: 8, 369: 8, 381: 5, 384: 8, 386: 8, 388: 8, 393: 7, 398: 8, 407: 7, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 455: 7, 456: 8, 458: 8, 463: 3, 467: 7, 469: 8, 479: 3, 481: 7, 485 ...
```
```suggestion
    190: 6, 193: 8, 197: 8, 199: 4, 201: 8, 209: 7, 211: 2, 241: 6, 249: 8, 288: 5, 298: 8, 304: 1, 309: 8, 311: 8, 313: 8, 320: 3, 322: 7, 328: 1, 352: 5, 353: 3, 381: 6, 386: 8, 388: 8, 393: 7, 398: 8, 407: 7, 413: 8, 417: 7, 419: 1, 422: 4, 426: 7, 431: 8, 442: 8, 451: 8, 452: 8, 453: 6, 455: 7, 456: 8, 463: 3, 479: 3, 481: 7, 485: 8, 487: 8, 489: 8, 495: 4, 497: 8, 499: 3, 500: 6, 501: 8, 503: 1, 508: 8, 510: 8, 528: 5, 532: 6, 534: 2, 554: 3, 560: 8, 562: 8, 563: 5, 564: 5, 565: 5, 567: 5, 573: 1, 608: 8, 609: 6, 610: 6, 611: 6, 612: 8, 613: 8, 647: 5, 707: 8, 753: 5, 761: 7, 801: 8, 804: 3, 810: 8, 840: 5, 842: 5, 844: 8, 866: 4, 872: 1, 882: 8, 890: 1, 892: 2, 893: 1, 894: 1, 961: 8, 967: 4, 969: 8, 977: 8, 979: 8, 985: 5, 1001: 8, 1005: 6, 1009: 8, 1011: 6, 1013: 3, 1017: 8, 1019: 2, 1020: 8, 1022: 1, 1105: 6, 1217: 8, 1221: 5, 1223: 2, 1225: 7, 1233: 8, 1243: 3, 1249: 8, 1257: 6, 1259: 8, 1261: 7, 1263: 4, 1265: 8, 1267: 1, 1280: 4, 1300: 8, 1322: 6, 1328: 4, 1417: 8, 1609: 8, 1613: 8, 1649: 8, 1792: 8, 1798: 8, 1824: 8, 1825: 8, 1840: 8, 1842: 8, 1858: 8, 1860: 8, 1863: 8, 1872: 8, 1875: 8, 1882: 8, 1888: 8, 1889: 8, 1892: 8, 1904: 7, 1906: 7, 1907: 7, 1912: 7, 1913: 7, 1914: 7, 1916: 7, 1918: 7, 1919: 7, 1937: 8, 1953: 8, 1968: 8, 2001: 8, 2017: 8, 2018: 8, 2020: 8, 2026: 8
```

koch-cf replied:
The new fingerprint works!  You’ll have to explain to me what you did, either here or on Discord.

sshane replied:
I fixed the fingerprint script here: https://github.com/commaai/openpilot/commit/48b059d3670e8082eae785c850508f88b1181188

Bus 0 and 2 are completely different on the GM ASCMs, so it was overriding some addresses with different messages entirely (and with different lengths as well).

Verylukyguy commented:
I think that the LaCrosse needs the Long Actuator Delay too.

sshane commented:
@Verylukyguy will keep that in mind. Let's do this one step at a time first though 🙂

koch-cf committed:
Update selfdrive/car/gm/values.py

Suggested fingerprint by @sshane

Co-authored-by: Shane Smiskol <shane@smiskol.com>

sshane committed:
Merge remote-tracking branch 'upstream/master' into koch-cf/master

sshane committed:
Premium is a trim with ACC. Use package name instead

twilsonco committed:
lacrosse custom FF;
fit info:
describe(steer_offsets) = DescribeResult(nobs=1649402, minmax=(-0.7127894163131714, 5.3997602462768555), mean=3.3090523060153645, variance=0.3130325564084465, skewness=-1.5986155151533736, kurtosis=8.18810418298873)
Samples: 1357787
Regularizing...
Regularized samples: 1140
speed: DescribeResult(nobs=1140, minmax=(8.478170424241286, 35.837870224662446), mean=27.888804767013475, variance=34.16742353763829, skewness=-1.068587303119431, kurtosis=0.6193071765927134)
angle: DescribeResult(nobs=1140, minmax=(-21.057768565637094, 28.516874490999708), mean=-0.4328602593886506, variance=43.19046813273241, skewness=0.001966426701503317, kurtosis=0.20547357649038434)
steer: DescribeResult(nobs=1140, minmax=(-0.8432471203007578, 0.9634959333674695), mean=-0.021981142946747863, variance=0.20152217060233915, skewness=0.03202313890158864, kurtosis=-1.2318826088567174)
Performing fit...
Fit: [5.85397825e-01 3.27650818e-01 4.60531117e-03 1.32307599e+01
 1.37194709e-01 1.33099557e-01 6.14782304e-02]
ANGLE_COEF = 0.58539783
ANGLE_COEF2 = 0.32765082
ANGLE_OFFSET = 0.00460531
SPEED_OFFSET = 13.23075991
SIGMOID_COEF_RIGHT = 0.13719471
SIGMOID_COEF_LEFT = 0.13309956
SPEED_COEF = 0.06147823
MAE old 0.2098, new 0.0309
STD old 0.1021, new 0.0273
deg 00-03:457, deg 03-06:258, deg 06-09:218, deg 09-12:132, deg 12-15:62
deg 15-18:6, deg 18-21:4, deg 21-24:1, deg 24-27:0, deg 27-30:2
deg 30-33:0, deg 33-36:0, deg 36-39:0, deg 39-42:0, deg 42-45:0

mph 10-15:0, mph 15-20:1, mph 20-25:18, mph 25-30:12, mph 30-35:30
mph 35-40:26, mph 40-45:52, mph 45-50:54, mph 50-55:73, mph 55-60:86
mph 60-65:204, mph 65-70:228, mph 70-75:179, mph 75-80:176, mph 80-85:1
mph 85-90:0,

koch-cf committed:
Merge pull request #1 from twilsonco/pr-lacrosse

lacrosse custom FF;

koch-cf commented on selfdrive/car/gm/interface.py:
```diff
@@ -36,12 +36,27 @@ def get_steer_feedforward_acadia(desired_angle, v_ego):
     return 0.04689655 * sigmoid * (v_ego + 10.028217)
+  
+  @staticmethod
+  def get_steer_feedforward_lacrosse(angle, speed):
+    ANGLE_COEF = 0.58539783
+    ANGLE_COEF2 = 0.32765082
+    ANGLE_OFFSET = 0.00460531
+    SPEED_OFFSET = 13.23075991
+    SIGMOID_COEF_RIGHT = 0.13719471
+    SIGMOID_COEF_LEFT = 0.13309956
+    SPEED_COEF = 0.06147823
+    x = ANGLE_COEF * (angle + ANGLE_OFFSET)
+    sigmoid = x / (1. + fabs(x))
+    return ((SIGMOID_COEF_RIGHT if (angle + ANGLE_OFFSET) > 0. else SIGMOID_COEF_LEFT) * sigmoid) * ((speed + SPEED_OFFSET) * SPEED_COEF) * ((fabs(angle + ANGLE_OFFSET) ** fabs(ANGLE_COEF2)))
```
My new C3 won’t arrive until tomorrow and I’ll be on a business trip until Friday, so next weekend will be the earliest that I can attempt a test drive. Hopefully the radar faults will be gone with the new hardware. 

Regarding the steering, I would agree with you in principle but in reality the stock tuning is borderline unusable. Below 40 mph it will oscillate dangerously in corners, and even on the highway it will touch the lines on both sides of the lane. You can see this reflected in the StDev measurements noted in the tune comments. So I would prefer to leave the custom tuning in, but would like your perspective on whether it’s better to include it now or in a separate PR.

sshane replied:
@nuwandavek do you have any opinion on merging temporary fixes for GM if it's this bad? Or should we wait for a general solution?

@koch-cf yes, let's make it a new PR since this sounds like it might block merging car support.

koch-cf replied:
Ran a test route with the custom lateral tuning and it is 100% better.  Car is now stable as low as 25mph, whereas with the stock tuning the car would literally weave back and forth across the entire road at that speed.  Did not get a chance yet to test high-speed instability (oscillating in sweeping turns at 65mph) but I predict that's also resolved.  If the custom tuning isn't acceptable for porting the car then something else is needed or it's going to be hazardous to drive.

koch-cf committed:
Update routes.py

sshane committed:
Merge remote-tracking branch 'upstream/master' into koch-cf/master

sshane committed:
Merge remote-tracking branch 'upstream/master' into koch-cf/master

sshane committed:
remove from non tested routes

sshane committed:
use torque controller

sshane committed:
update docs

sshane committed:
update releases

zlg810 commented:
May I ask if this is compatible with the 2021 Buick LaCrosse?

zhou8603051 commented:
请问这是否与 2021 款别克君越兼容
